src/cli/subcommand/export.rs

```diff
@@ -0,0 +1,270 @@
+use std::collections::HashMap;
+use std::env;
+use std::ffi::{OsStr, OsString};
+use std::io::{stdout, Write};
+use std::os::unix::ffi::OsStrExt;
+use std::path::PathBuf;
+use std::process::ExitCode;
+
+use crate::cli::CommandExecute;
+use clap::Parser;
+
+const LOCAL_STATE_DIR: &str = "/nix/var";
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("The HOME environment variable is not set.")]
+    HomeNotSet,
+
+    #[error("__ETC_PROFILE_NIX_SOURCED is set, indicating the relevant environment variables have already been set.")]
+    AlreadyRun,
+
+    #[error("Some of the paths from Nix for XDG_DATA_DIR are not valid, due to an illegal character, like a colon.")]
+    InvalidXdgDataDirs(Vec<PathBuf>),
+
+    #[error("Some of the paths from Nix for PATH are not valid, due to an illegal character, like a colon.")]
+    InvalidPathDirs(Vec<PathBuf>),
+
+    #[error("Some of the paths from Nix for MANPATH are not valid, due to an illegal character, like a colon.")]
+    InvalidManPathDirs(Vec<PathBuf>),
+}
+
+/**
+Emit all the environment variables that should be set to use Nix.
+
+Safety note: environment variables and values can contain any bytes except
+for a null byte. This includes newlines and spaces, which requires careful
+handling.
+
+In `space-newline-separated` mode, `nix-installer` guarantees it will:
+
+  * only emit keys that are alphanumeric with underscores,
+  * only emit values without newlines
+
+and will refuse to emit any output to stdout if the variables and values
+would violate these safety rules.
+
+In `null-separated` mode, `nix-installer` emits data in this format:
+
+  KEYNAME\0VALUE\0KEYNAME\0VALUE\0
+
+*/
+#[derive(Debug, Parser)]
+#[command(args_conflicts_with_subcommands = true)]
+pub struct Export {
+    #[clap(long)]
+    format: ExportFormat,
+
+    #[clap(long)]
+    sample_output: Option<PathBuf>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, clap::ValueEnum)]
+enum ExportFormat {
+    Fish,
+    Sh,
+}
+
+#[async_trait::async_trait]
+impl CommandExecute for Export {
+    #[tracing::instrument(level = "trace", skip_all)]
+    async fn execute(self) -> eyre::Result<ExitCode> {
+        let env: HashMap<String, OsString> = match self.sample_output {
+            Some(filename) => {
+                // Note: not tokio File b/c I don't think serde_json has fancy async support?
+                let file = std::fs::File::open(filename)?;
+                let intermediate: HashMap<String, String> = serde_json::from_reader(file)?;
+                intermediate
+                    .into_iter()
+                    .map(|(k, v)| (k, v.into()))
+                    .collect()
+            },
+            None => {
+                match calculate_environment() {
+                    e @ Err(Error::AlreadyRun) => {
```

Comment:
I'm not sure this is ideal. Returning errors invokes various backtrace handling which could add several microseconds to runtime. 

Consider checking the env var on a new line between L82 and L83:

```rust

    if nonempty_var_os("__ETC_PROFILE_NIX_SOURCED") == Some("1".into()) {
         tracing::debug!("Already ran export");
         return Ok(ExitCode::SUCCESS);
    }
```

Then removing that error variant from the error type.